src/api/cpp/nixl_v2_2.h

```diff
@@ -0,0 +1,125 @@
+
+    Agent(std::string_view name, const Config &cfg);
+    ~Agent();
+
+    ///////////////////////////////////////////////////////////////////
+    // Metadata / connection setup
+
+    // Maybe return remote agent name from metadata
+    std::string getMetadata();
+    void connect(const std::string &remote_metadata);
+
+
+    // TODO support storage get-closest APIs (Vish)
+    // TODO How callbacks/no callbacks will map to Rust API (Tim)
+
+    ///////////////////////////////////////////////////////////////////
+    // Prepare transfer list
+
+    // Prepare local and remote buffers for transfer:
+    // - Register local buffers
+    // - Potentially calculate the communication pattern and breakdown to
+    //   collective operations
+    // - Potentially send a message to the remote agent(s) asking to resolve the
+    //   remote addresses to memory access keys, or use one-sided RDMA_READ
+    //   to resolve remote keys.
+    // - List sizes and corresponding buffer sizes are the same.
+    // We considered add signal list / inline with remote dst, but for now we
+    // decided to only keep the separate message API.
+    XferList makeXferList(const std::vector<MemDesc> &dst_list,
+                          const std::vector<MemDesc> &src_list,
+                          const OptionalParams &params);
+
+    // Estimate the time to complete the transfer, in seconds.
+    // In case of storage it uses the storage SLA from vendor SDK.
+    double estimateXferTime(const XferList &xfer_list);
+
+    ///////////////////////////////////////////////////////////////////
+    // Posting list of operations
+
+    // Post all operations in the list
+    // There is no ordering guarantee between different transfer requests or
+    // between operations in the same transfer request.
+    XferRequest postXfer(const XferList &xfer_list);
+
+    // Post operations in the list with specific dst and src indices
+    XferRequest postXfer(const XferList &xfer_list,
+                         const std::vector<int> &dst_indices,
+                         const std::vector<int> &src_indices);
+
+    ///////////////////////////////////////////////////////////////////
+    // Completion checking
+
+    // Poll Option1: check if a specific transfer is done
+    // TODO: allow returning failures
+    // TODO move it to XferRequest class?
+    bool isXferDone(const XferRequest &xfer_request);
+
+    // Poll Option2: get all completed transfers
+    void getCompletedXfers(std::vector<XferRequest> &xfer_requests);
+
+    ///////////////////////////////////////////////////////////////////
+    // Signals / Notifications
+
+    // Post a signal to a remote agent
+    // Signal is ordered w.r.t. to any previous transfers to the same remote
+    // agent.
+    void postMessage(std::string_view agent_name, const std::string &message);
+
+    // Option1: poll for all signals
+    void pollMessages(std::multimap<std::string_view, std::string> &signals);
```

Comment:
Who owns the string that string_view points to in this case? The agent instance? How it the lifetime of the target is accounted for considering that caller can retain the resulting multimap indefinitely. 

PS also, return the container instead of modifying by reference.

Reply:
Yes, the agent owns the string view (since the first element is the name of the remote agent, and wanted to avoid extra string copy)

Reply:
So if, for example, agent->invalidateRemoteMD() (using a method from existing API to simplify) is called concurrently user would suddenly end up with dangling pointer to a deallocated string? Just to save a string copy... when already constructing a std::multimap instance?